python-libraries/data-platform-catalogue/data_platform_catalogue/client.py

```diff
@@ -65,9 +121,55 @@ class ReferencedEntityMissing(CatalogueError):
     """
 
 
-class CatalogueClient:
+class BaseCatalogueClient(ABC):
+    @abstractmethod
+    def upsert_database_service(
+        self, *args: Any, **kwargs: Any
+    ):  # type: ignore[override]
+        pass
+
+    @abstractmethod
+    def upsert_database(self, *args: Any, **kwargs: Any):  # type: ignore[override]
+        pass
+
+    @abstractmethod
+    def upsert_schema(self, *args: Any, **kwargs: Any):  # type: ignore[override]
+        pass
+
+    @abstractmethod
+    def upsert_table(
```

Comment:
I don't think the use of `*args` and `**kwargs` is doing what you expect here.

Since this is an abstract base class, we would expect to be able to type any value to `BaseCatalogueClient` and substitute any of the concrete classes without breaking things. But this is not the case, since the base class accepts *any* arguments, and the concrete classes do not.

(See [Liskov's substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle))

I think what you want to do instead is write out the full arglist here, and then use *args, **kwargs in the subclasses only when you want to ignore some of them.